### DISCUSSION TITLE
Src URL uses set url generator, but srcset URLs do not?

### DISCUSSION THREAD

**Discussion by torbjorn-stabo** [spatie/laravel-medialibrary#3675] [Q&A]
Hi. I'm using spatie/laravel-medialibrary 11.7.6 and stancl/tenancy 3.8.4, and my goal here is to make them work together.

I've edited config/media-library.php:
```php
// 'url_generator' => Spatie\MediaLibrary\Support\UrlGenerator\DefaultUrlGenerator::class,
 'url_generator' => TenantAwareUrlGenerator::class,
``` 
 ..and the code for TenantAwareUrlGenerator looks like this:
```php
<?php
namespace App\Services;

use Spatie\LaravelSettings\Support;
use Spatie\MediaLibrary\Support\UrlGenerator\DefaultUrlGenerator;

class TenantAwareUrlGenerator extends DefaultUrlGenerator
{
    public function getUrl(): string
    {
        $url = tenant_asset($this->getPathRelativeToRoot());

        $url = $this->versionUrl($url);

        return $url;
    }
}
```  

When "fetching" the HTML for the media image I do this:
`$item->getFirstMedia('mymediacollection')?->toHtml();`
..where $item is an instance of a class with the default medialibrary Media linked to it in the database.

Problem is, the src in the `<img>` is tenant aware, i.e the URL is `http://mysite.localhost/...` which is correct. In the srcset the URLs are `http://localhost/...`.

Stepping through things I see this:
`responsiveImageWithPlaceholder.blade.php` contains this:
```php
<img
  <?php
  srcset="<?php echo e($media->getSrcset($conversion)); ?>"
  src="<?php echo e($media->getUrl($conversion)); ?>"
  ...
>

``` 
`$media->getUrl()` uses the tenant aware Url Generator configured earlier, but `$media->getSrcset($conversion)` doesn't.

If needed I'll of course paste any extra info needed. Tbh I'm not sure whether to ask about this here or in the Tenancy for laravel forum, I hope I'm not offending anyone.

**Comment by torbjorn-stabo** [comment]
Additional info:
When loading the responsive images for `srcset` the code reaches `FileSystemAdapter::url()`, and since it's a local file getLocalUrl() is called, and that method contains this:
```php
        if (isset($this->config['url'])) {
            return $this->concatPathToUrl($this->config['url'], $path);
        }
``` 
`$this->config['url']` initially contains `'http://localhost/storage'`. If I unset it the code below that if() generates absolute paths, e.g `srcset="/storage/responsive-images/myimg.png ..."`.

`http://localhost/storage` comes from the disk(public in my case) config in filesystems.php. The config contains `'url' => env('APP_URL').'/storage'`, and by default (I think) `env('APP_URL')=http://localhost` (static). But why is path generation for responsive images bypassing the url_generator configured in media-library.php? We're talking filenames in both cases, with paths built and prepended, right?

**Comment by torbjorn-stabo** [comment]
Are src and srcset using different path generators? I'm running out of ideas on this one.

**Comment by AhmedSobhy01** [comment]
The directory URL for responsive images is generated using `getResponsiveImagesDirectoryUrl` defined on the `UrlGenerator` interface.

The solution would be to override this function too in your `TenantAwareUrlGenerator` class so it becomes like this:
```php
<?php
namespace App\Services;

use Spatie\LaravelSettings\Support;
use Spatie\MediaLibrary\Support\UrlGenerator\DefaultUrlGenerator;
use Illuminate\Support\Str;

class TenantAwareUrlGenerator extends DefaultUrlGenerator
{
    public function getUrl(): string
    {
        $url = tenant_asset($this->getPathRelativeToRoot());

        $url = $this->versionUrl($url);

        return $url;
    }

    public function getResponsiveImagesDirectoryUrl(): string
    {
        $path = $this->pathGenerator->getPathForResponsiveImages($this->media);

        return Str::finish(tenant_asset($path), '/');
    }
}
```
